text/0000-otlp-http.md

```diff
@@ -0,0 +1,258 @@
+overloaded the server SHOULD respond with `HTTP 429 Too Many Requests` or
+`HTTP 503 Service Unavailable` and MAY include "Retry-After" header with a
+recommended time interval in seconds to wait before retrying. The client SHOULD
+honour the waiting interval specified in "Retry-After" header if it is present.
+If the client receives `HTTP 429` or `HTTP 503` response and "Retry-After"
+header is not present in the response then the client SHOULD implement an
+exponential backoff strategy between retries.
+
+#### All Other Responses
+
+All other HTTP responses that are not explicitly listed in this document should
+be treated according to HTTP specification.
+
+If the server disconnects without returning a response the client SHOULD retry
+and send the same request. The client SHOULD implement an exponential backoff
+strategy between retries to avoid overwhelming the server.
+
+### Connection
+
+If the client is unable to connect to the server the client SHOULD retry the
+connection using exponential backoff strategy between retries. The interval
+between retries must have a random jitter.
+
+The client SHOULD keep the connection alive between requests. The client SHOULD
+use "Connection: Keep-Alive" request header.
+
+Server implementations MAY handle OTLP/gRPC and OTLP/HTTP requests on the same
+port and multiplex the connections to the corresponding transport handler based
+on "Content-Type" request header.
+
+### Parallel Connections
+
+To achieve higher total throughput the client MAY send requests using several
+parallel HTTP connections. In that case the maximum number of parallel
+connections SHOULD be configurable.
+
+## Prior Art and Alternatives
+
+I have also considered HTTP/1.1+WebSocket transport. Experimental implementation
+of OTLP over WebSocket transport has shown that it typically has better
+performance than plain HTTP transport implementation (WebSocket uses less CPU,
+higher throughput in high latency connections). However WebSocket transport
+requires slightly more complicated implementation and WebSocket libraries are
+less ubiquitous than plain HTTP, which may make implementation in certain
+languages difficult or impossible.
+
+HTTP/1.1+WebSocket transport may be considered as a future transport for
+high-performance use cases as it exhibits better performance than OTLP/gRPC and
+OTLP/HTTP.
+
+## Performance Test Results
+
+The following is the benchmarking result of
+[Go implementations](https://github.com/tigrannajaryan/exp-otelproto/tree/master/cmd/benchmark),
+running on on Ubuntu 18 LTS system with i7 7500U processor, 16 GB RAM. The test
+benchmarks loopback sending and receiving of span batches. (Note that the
+benchmarking script sets "performance" CPU governor during execution and sets
+nice value of the process for more consistent results).
+
+```
+====================================================================================
+Legend:
```

Comment:
It is not necessary a correct comparison. There are some features like deadline/retries/etc. that gRPC has and I don't know if they are present in the benchmarks for HTTP.

Reply:
OK, let me remove the benchmark since it does not really matter for the motivation of this proposal.